runner/src/main.rs

```diff
@@ -331,40 +342,54 @@ struct Client {
+fn run_example(opt: &RunExamples, example: &Example) -> Step {
     Step::Multiple {
         name: example.name.to_string(),
         steps: vec![
-            Step::Multiple {
-                name: "build_wasm_modules".to_string(),
-                steps: example
-                    .modules
-                    .iter()
-                    .map(|(name, target)| match target {
-                        Target::Cargo { cargo_manifest } => {
-                            build_wasm_module(name, &cargo_manifest)
-                        }
-                        Target::Bazel { .. } => todo!(),
-                    })
-                    .collect(),
-            },
-            build_example_config(&example.name),
-            Step::WithBackground {
-                name: "run_server".to_string(),
-                background: run_example_server(&format!(
-                    "./bazel-bin/examples/{}/config/config.bin",
-                    example.name
-                )),
-                foreground: Box::new(Step::Multiple {
-                    name: "run_client".to_string(),
+            vec![
+                Step::Multiple {
+                    name: "build_wasm_modules".to_string(),
                     steps: example
-                        .clients
+                        .modules
                         .iter()
-                        .map(|(name, client)| run_client(name, &client))
+                        .map(|(name, target)| match target {
+                            Target::Cargo { cargo_manifest } => {
+                                build_wasm_module(name, &cargo_manifest)
+                            }
+                            Target::Bazel { .. } => todo!(),
+                        })
                         .collect(),
+                },
+                build_example_config(&example.name),
+                // Build the server first so that when running it in the next step it will start up
+                // faster.
+                build_server(&BuildServer {
+                    variant: "rust".to_string(),
                 }),
+            ],
+            if opt.build_only {
+                vec![]
+            } else {
+                vec![Step::WithBackground {
+                    name: "run_server".to_string(),
+                    background: run_example_server(&format!(
+                        "./bazel-bin/examples/{}/config/config.bin",
+                        example.name
+                    )),
+                    foreground: Box::new(Step::Multiple {
+                        name: "run_client".to_string(),
+                        steps: example
+                            .clients
+                            .iter()
+                            .map(|(name, client)| run_client(name, &client))
+                            .collect(),
+                    }),
+                }]
             },
-        ],
+        ]
+        .into_iter()
+        .flatten()
+        .collect::<Vec<_>>(),
```

Comment:
It interesting to see assembling a Vec based on conditions is done in Rust.

In JS land, the spread operator makes that very easy.

Reply:
Yeah this is quite verbose, but this is effectively allocating a new vec composed of those parts, and Rust always makes allocations very explicit.

Reply:
Yeah, I figured that was the advantage of that. :)